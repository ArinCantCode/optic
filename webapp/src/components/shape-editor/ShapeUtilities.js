import { boundParameterColor, primitiveColors } from './Types.js';

export const coreShapeIds = ['$string', '$number', '$boolean', '$object', '$list', '$map', /*'$oneOf',*/ '$identifier', '$reference', '$any', '$nullable', '$optional', '$unknown']
export const coreShapeIdsSet = new Set(coreShapeIds)

class ShapeUtilities {
    static flatten(queries, rootShapeId, depth, trail = [], acc = []) {
        const shape = queries.shapeById(rootShapeId)
        const { name, baseShapeId, bindings, fields } = shape;
        if (coreShapeIdsSet.has(rootShapeId)) {
            return;
        }
        const baseShape = queries.shapeById(baseShapeId)
        const shouldTraverseFields = ShapeUtilities.shouldTraverseFields(queries, shape)
        if (trail.length === 0 || name !== '') {
            acc.push({
                id: shape.shapeId,
                type: 'shape',
                name: name,
                shapeName: ShapeUtilities.shapeName(queries, baseShape, shape, bindings),
                isExpandable: shouldTraverseFields,
                trail: [...trail, rootShapeId],
                depth
            })
        }

        if (shouldTraverseFields) {
            const parentObject = ShapeUtilities.resolveObjectWithFields(queries, shape)
            const { fields } = parentObject;
            fields.forEach(field => {
                if (field.isRemoved) {
                    return
                }
                const fieldTrail = [...trail, rootShapeId, field.fieldId]
                const [isInlineObject, fieldShape] = ShapeUtilities.isInlineObject(queries, field)
                acc.push({
                    id: field.fieldId,
                    parentShapeId: parentObject.shapeId,
                    fieldShapeId: fieldShape ? fieldShape.shapeId : null,
                    type: 'field',
                    name: field.name,
                    shapeName: ShapeUtilities.fieldShapeName(queries, shape, field, bindings),
                    trail: fieldTrail,
                    isExpandable: isInlineObject,
                    depth: depth + 1
                })
                if (isInlineObject) {
                    ShapeUtilities.flatten(queries, fieldShape.shapeId, depth + 1, fieldTrail, acc)
                }
            })

        }
    }

    static shouldTraverseFields(queries, shape) {
        // console.log({ shape })
        return shape.coreShapeId === '$object'
    }

    static resolveObjectWithFields(queries, shape) {
        if (shape.coreShapeId !== '$object') {
            throw new Error(`can't resolve fields of a non-object`)
        }
        let resolvedShape = shape;
        while (resolvedShape.baseShapeId !== "$object") {
            resolvedShape = queries.shapeById(resolvedShape.baseShapeId)
        }
        return resolvedShape
    }

    static isInlineObject(queries, field) {
        const { fieldShapeDescriptor } = field

        if (fieldShapeDescriptor.FieldShapeFromShape) {
            const shape = queries.shapeById(fieldShapeDescriptor.FieldShapeFromShape.shapeId)
            const name = shape.name
            if (shape.baseShapeId === '$object' && name === '') {
                return [true, shape]
            }
            return [false, shape]
        } else if (fieldShapeDescriptor.FieldShapeFromParameter) {
            return [false]
        }

    }

    static mergeBindings(parentObjectBindings, fieldBindings) {
        const keySet = new Set([parentObjectBindings, fieldBindings].map(x => Object.keys(x)).flatMap(x => x))
        // console.log({output: keySet})
        const merged = [...keySet.values()].map(x => [x, fieldBindings[x] || parentObjectBindings[x]]).reduce((acc, item) => {
            const [key, value] = item
            acc[key] = value
            return acc;
        }, {})
        // console.log({output: merged})
        return merged
    }

    static fieldShapeName(queries, shape, field, parentBindings) {
        const { fieldShapeDescriptor } = field;
        if (fieldShapeDescriptor.FieldShapeFromShape) {
            const fieldShapeId = fieldShapeDescriptor.FieldShapeFromShape.shapeId
            const fieldShape = queries.shapeById(fieldShapeId)
            return ShapeUtilities.shapeName(queries, fieldShape, shape, ShapeUtilities.mergeBindings(parentBindings, field.bindings))
        } else if (fieldShapeDescriptor.FieldShapeFromParameter) {
            const parameterId = fieldShapeDescriptor.FieldShapeFromParameter.shapeParameterId
            const parameter = shape.parameters.find(x => x.shapeParameterId === parameterId)
            if (!parameter) {
                console.warn(`expected parameter ${parameterId} to be in parent shape ${shape.shapeId} for field ${field.fieldId}`)
            }
            return {
                parameterName: parameter.name,
                parameterId: parameter.shapeParameterId,
                bindingInfo: [],
            }
        }
    }

    static shapeName(queries, baseShape, bindingContextShape, bindings) {
        const baseShapeOfBaseShape = queries.shapeById(baseShape.baseShapeId)
        const base = {
            id: baseShape.shapeId,
            color: primitiveColors[baseShape.baseShapeId],
            coreShapeId: baseShape.coreShapeId,
            userDefinedName: baseShape.name,
            baseShapeName: baseShape.name || baseShapeOfBaseShape.name,
        }
        if (baseShape.parameters.length === 0 && baseShapeOfBaseShape.parameters.length === 0) {
            return {
                ...base,
                bindingInfo: []
            }
        }
        const parameters = [...baseShape.parameters]
        const parameterIds = new Set(parameters.map(x => x.parameterId))
        baseShapeOfBaseShape.parameters.forEach((x => {
            if (!parameterIds.has(x.parameterId)) {
                parameters.push(x)
            }
        }))
        const bindingNames = parameters
            .map(parameter => {
                const binding = bindings[parameter.shapeParameterId]
                if (binding) {
                    if (binding.ShapeProvider) {
                        const { shapeId } = binding.ShapeProvider;
                        const shape = queries.shapeById(shapeId)
                        return {
                            binding,
                            color: primitiveColors[shape.baseShapeId],
                            boundName: queries.shapeById(shapeId).name || queries.shapeById(shape.baseShapeId).name || '(unnamed)',
                            parameterName: parameter.name,
                            parameterId: parameter.shapeParameterId,
                        }
                    }

                    if (binding.ParameterProvider) {
                        const parameterId = binding.ParameterProvider.shapeParameterId
                        const provider = bindingContextShape.parameters.find(x => x.shapeParameterId === parameterId)
                        if (!provider) {
                            console.warn(`expected parameter ${parameterId} to be in parent shape ${bindingContextShape.shapeId}`)
                        }
                        return {
                            binding,
                            color: boundParameterColor,
                            boundName: provider ? provider.name || '(unnamed)' : parameter.name, //@TODO resolve parent bindings
                            parameterName: parameter.name,
                            parameterId: parameter.shapeParameterId,
                        }
                    }
                }
                return {
                    parameterName: parameter.name,
                    parameterId: parameter.shapeParameterId,
                }
            })
        return {
            ...base,
            bindingInfo: bindingNames
        }
    }
}

export {
    ShapeUtilities
}
