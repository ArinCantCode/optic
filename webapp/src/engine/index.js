import * as seamlessEngine from 'seamless-domain'

export const ShapesCommands = seamlessEngine.com.seamless.contexts.shapes.Commands
export const ShapesHelper = seamlessEngine.com.seamless.contexts.shapes.ShapesHelper()
export const RequestsHelper = seamlessEngine.com.seamless.contexts.requests.RequestsServiceHelper()
export const ContentTypesHelper = seamlessEngine.com.seamless.contexts.requests.ContentTypes()

export const RfcCommands = seamlessEngine.com.seamless.contexts.rfc.Commands
export const RequestsCommands = seamlessEngine.com.seamless.contexts.requests.Commands

export const Facade = seamlessEngine.com.seamless.contexts.rfc.RfcServiceJSFacade()
export const Queries = (eventStore, service, aggregateId) => new seamlessEngine.Queries(eventStore, service, aggregateId)


export function commandsToJson(commands) {
    return commands.map(x => JSON.parse(seamlessEngine.CommandSerialization.toJsonString(x)))
}
export function commandsFromJson(commands) {
    return seamlessEngine.CommandSerialization.fromJsonString(JSON.stringify(commands))
}

const { ApiInteraction, ApiRequest, ApiResponse } = seamlessEngine.com.seamless.diff;
export const JsonHelper = seamlessEngine.com.seamless.diff.JsonHelper()
function fromJs(x) {
    return JsonHelper.fromString(JSON.stringify(x))
}

export function toInteraction(sample) {
    return ApiInteraction(
        ApiRequest(sample.request.url, sample.request.method, sample.request.headers['content-type'] || '*/*', fromJs(sample.request.body)),
        ApiResponse(sample.response.statusCode, sample.response.headers['content-type'] || '*/*', fromJs(sample.response.body))
    )
}
export const RequestDiffer = seamlessEngine.com.seamless.diff.RequestDiffer()
export const DiffToCommands = seamlessEngine.com.seamless.diff.DiffToCommands()
console.log(seamlessEngine)