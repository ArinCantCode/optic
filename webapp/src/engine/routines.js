import {RfcCommands} from './index';

export function updateContribution(id, key, value) {
	return RfcCommands.AddContribution(id, key, value)
}

export function renameAPI(newName) {
	return RfcCommands.SetAPIName(newName)
}
