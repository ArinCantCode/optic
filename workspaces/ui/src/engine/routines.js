import {RfcCommands, ShapesCommands} from '@useoptic/domain';

export function updateContribution(id, key, value) {
	return RfcCommands.AddContribution(id, key, value)
}

export function renameShape(shapeId, name) {
	return ShapesCommands.RenameShape(shapeId, name)
}

export function renameAPI(newName) {
	return RfcCommands.SetAPIName(newName)
}

