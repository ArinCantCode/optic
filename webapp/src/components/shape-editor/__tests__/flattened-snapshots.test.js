import { Facade, Queries, ShapesCommands, RfcCommandContext } from '../../../engine';
import { ShapeUtilities } from '../ShapeUtilities.js';

const starterCommands = [
    ShapesCommands.AddShape('$Id', '$identifier', 'ID'),
    ShapesCommands.SetParameterShape(ShapesCommands.ProviderInShape('$Id', ShapesCommands.ShapeProvider('$string'), '$identifierInner')),

    ShapesCommands.AddShape('$Account', '$object', 'Account'),
    ShapesCommands.AddShape('$AccountId', '$reference', 'Account ID'),
    ShapesCommands.SetParameterShape(ShapesCommands.ProviderInShape('$AccountId', ShapesCommands.ShapeProvider('$Account'), '$referenceInner')),

    ShapesCommands.AddShape('$User', '$object', 'User'),
    ShapesCommands.AddShape('$UserId', '$reference', 'User ID'),
    ShapesCommands.SetParameterShape(ShapesCommands.ProviderInShape('$UserId', ShapesCommands.ShapeProvider('$User'), '$referenceInner')),

    ShapesCommands.AddField('$Account.id', '$Account', 'id', ShapesCommands.FieldShapeFromShape('$Account.id', '$AccountId')),
    ShapesCommands.AddField('$Account.userIds', '$Account', 'userIds', ShapesCommands.FieldShapeFromShape('$Account.userIds', '$list')),
    ShapesCommands.SetParameterShape(ShapesCommands.ProviderInField('$Account.userIds', ShapesCommands.ShapeProvider('$UserId'), '$listItem')),

    ShapesCommands.AddField('$User.id', '$User', 'id', ShapesCommands.FieldShapeFromShape('$User.id', '$string')),
    ShapesCommands.AddField('$User.name', '$User', 'name', ShapesCommands.FieldShapeFromShape('$User.name', '$string')),
    ShapesCommands.AddField('$User.accountId', '$User', 'accountId', ShapesCommands.FieldShapeFromShape('$User.accountId', '$AccountId')),

    ShapesCommands.AddShape('$PaginationWrapper', '$object', 'PaginationWrapper'),
    ShapesCommands.AddShapeParameter('$PaginationWrapper.T', '$PaginationWrapper', 'T'),
    ShapesCommands.AddField('$PaginationWrapper.offset', '$PaginationWrapper', 'offset', ShapesCommands.FieldShapeFromShape('$PaginationWrapper.offset', '$number')),
    ShapesCommands.AddField('$PaginationWrapper.items', '$PaginationWrapper', 'items', ShapesCommands.FieldShapeFromShape('$PaginationWrapper.items', '$list')),
    ShapesCommands.SetParameterShape(ShapesCommands.ProviderInField('$PaginationWrapper.items', ShapesCommands.ParameterProvider('$PaginationWrapper.T'), '$listItem')),

    ShapesCommands.AddShape('$UserListResponse1', '$PaginationWrapper', 'User List Response 1'),
    ShapesCommands.SetParameterShape(ShapesCommands.ProviderInShape('$UserListResponse1', ShapesCommands.ShapeProvider('$User'), '$PaginationWrapper.T')),

    ShapesCommands.AddShape('$UserListResponse2', '$PaginationWrapper', 'User List Response 2'),

    ShapesCommands.AddShape('$UserListResponse3', '$PaginationWrapper', 'User List Response 3'),
    ShapesCommands.AddShapeParameter('$UserListResponse3.X', '$UserListResponse3', 'X'),
    ShapesCommands.AddShapeParameter('$UserListResponse3.Y', '$UserListResponse3', 'Y'),
    ShapesCommands.SetParameterShape(ShapesCommands.ProviderInShape('$UserListResponse3', ShapesCommands.ParameterProvider('$UserListResponse3.X'), '$PaginationWrapper.T')),
    ShapesCommands.AddField('$UserListResponse3.item', '$UserListResponse3', 'item', ShapesCommands.FieldShapeFromParameter('$UserListResponse3.item', '$UserListResponse3.Y'))
]

const todoEvents = [
    { "PathComponentAdded": { "name": "todo", "parentPathId": "root", "pathId": "path_ddceX9ubhV" } }
    , { "PathParameterAdded": { "name": "todoId", "parentPathId": "path_ddceX9ubhV", "pathId": "path_z69omkQCeH" } }
    , { "ShapeAdded": { "baseShapeId": "$string", "name": "", "parameters": { "DynamicParameterList": { "shapeParameterIds": [] } }, "shapeId": "shape_W4mBGJW3Ex" } }
    , { "PathParameterShapeSet": { "pathId": "path_z69omkQCeH", "shapeDescriptor": { "isRemoved": false, "shapeId": "shape_W4mBGJW3Ex" } } }
    , { "RequestAdded": { "httpMethod": "PATCH", "pathId": "path_z69omkQCeH", "requestId": "request_ZkPTjQXkYh" } }
    , { "ShapeAdded": { "baseShapeId": "$object", "name": "", "parameters": { "DynamicParameterList": { "shapeParameterIds": [] } }, "shapeId": "9IRsry_0" } }
    , { "ShapeAdded": { "baseShapeId": "$boolean", "name": "", "parameters": { "DynamicParameterList": { "shapeParameterIds": [] } }, "shapeId": "9IRsry_2" } }
    , { "FieldAdded": { "fieldId": "9IRsry_1", "name": "isDone", "shapeDescriptor": { "FieldShapeFromShape": { "fieldId": "9IRsry_1", "shapeId": "9IRsry_2" } }, "shapeId": "9IRsry_0" } }
    , { "ShapeAdded": { "baseShapeId": "$string", "name": "", "parameters": { "DynamicParameterList": { "shapeParameterIds": [] } }, "shapeId": "9IRsry_4" } }
    , { "FieldAdded": { "fieldId": "9IRsry_3", "name": "task", "shapeDescriptor": { "FieldShapeFromShape": { "fieldId": "9IRsry_3", "shapeId": "9IRsry_4" } }, "shapeId": "9IRsry_0" } }
    , { "ShapeAdded": { "baseShapeId": "9IRsry_0", "name": "", "parameters": { "DynamicParameterList": { "shapeParameterIds": [] } }, "shapeId": "shape_IrEV2HYW8k" } }
    , { "RequestBodySet": { "bodyDescriptor": { "httpContentType": "application/json", "isRemoved": false, "shapeId": "shape_IrEV2HYW8k" }, "requestId": "request_ZkPTjQXkYh" } }
    , { "ShapeRenamed": { "name": "ToDoUpdateBody", "shapeId": "9IRsry_0" } }
    , { "ResponseAdded": { "httpStatusCode": 200, "requestId": "request_ZkPTjQXkYh", "responseId": "response_IdpJexoAGm" } }
    , { "ShapeAdded": { "baseShapeId": "$object", "name": "", "parameters": { "DynamicParameterList": { "shapeParameterIds": [] } }, "shapeId": "_Todo_0" } }
    , { "ShapeAdded": { "baseShapeId": "$string", "name": "", "parameters": { "DynamicParameterList": { "shapeParameterIds": [] } }, "shapeId": "_Todo_2" } }
    , { "FieldAdded": { "fieldId": "_Todo_1", "name": "id", "shapeDescriptor": { "FieldShapeFromShape": { "fieldId": "_Todo_1", "shapeId": "_Todo_2" } }, "shapeId": "_Todo_0" } }
    , { "ShapeAdded": { "baseShapeId": "$boolean", "name": "", "parameters": { "DynamicParameterList": { "shapeParameterIds": [] } }, "shapeId": "_Todo_4" } }
    , { "FieldAdded": { "fieldId": "_Todo_3", "name": "isDone", "shapeDescriptor": { "FieldShapeFromShape": { "fieldId": "_Todo_3", "shapeId": "_Todo_4" } }, "shapeId": "_Todo_0" } }
    , { "ShapeAdded": { "baseShapeId": "$string", "name": "", "parameters": { "DynamicParameterList": { "shapeParameterIds": [] } }, "shapeId": "_Todo_6" } }
    , { "FieldAdded": { "fieldId": "_Todo_5", "name": "task", "shapeDescriptor": { "FieldShapeFromShape": { "fieldId": "_Todo_5", "shapeId": "_Todo_6" } }, "shapeId": "_Todo_0" } }
    , { "ShapeAdded": { "baseShapeId": "_Todo_0", "name": "", "parameters": { "DynamicParameterList": { "shapeParameterIds": [] } }, "shapeId": "shape_sS9qO1iLdD" } }
    , { "ResponseBodySet": { "bodyDescriptor": { "httpContentType": "application/json; charset=utf-8", "isRemoved": false, "shapeId": "shape_sS9qO1iLdD" }, "responseId": "response_IdpJexoAGm" } }
    , { "ShapeRenamed": { "name": "ToDo", "shapeId": "_Todo_0" } }
    , { "ContributionAdded": { "id": "request_ZkPTjQXkYh", "key": "description", "value": "Update the Todo in the path (todoId)" } }
    , { "PathComponentAdded": { "name": "todos", "parentPathId": "root", "pathId": "path_psf1AdDPdC" } }
    , { "RequestAdded": { "httpMethod": "GET", "pathId": "path_psf1AdDPdC", "requestId": "request_jyqcOCtAdq" } }
    , { "ResponseAdded": { "httpStatusCode": 200, "requestId": "request_jyqcOCtAdq", "responseId": "response_m0cAG6CQAY" } }
    , { "ShapeAdded": { "baseShapeId": "$list", "name": "", "parameters": { "DynamicParameterList": { "shapeParameterIds": [] } }, "shapeId": "rG6Eyi_0" } }
    , { "ShapeParameterShapeSet": { "shapeDescriptor": { "ProviderInShape": { "consumingParameterId": "$listItem", "providerDescriptor": { "ShapeProvider": { "shapeId": "_Todo_0" } }, "shapeId": "rG6Eyi_0" } } } }
    , { "ShapeAdded": { "baseShapeId": "rG6Eyi_0", "name": "", "parameters": { "DynamicParameterList": { "shapeParameterIds": [] } }, "shapeId": "_ResponseBody_0" } }
    , { "ResponseBodySet": { "bodyDescriptor": { "httpContentType": "application/json; charset=utf-8", "isRemoved": false, "shapeId": "_ResponseBody_0" }, "responseId": "response_m0cAG6CQAY" } }
    , { "ContributionAdded": { "id": "request_jyqcOCtAdq", "key": "description", "value": "Get a list of all the todos" } }
    , { "RequestAdded": { "httpMethod": "POST", "pathId": "path_psf1AdDPdC", "requestId": "request_8VF0hXMHz0" } }
    , { "ShapeAdded": { "baseShapeId": "9IRsry_0", "name": "", "parameters": { "DynamicParameterList": { "shapeParameterIds": [] } }, "shapeId": "shape_f4ymchjzZV" } }
    , { "RequestBodySet": { "bodyDescriptor": { "httpContentType": "application/json", "isRemoved": false, "shapeId": "shape_f4ymchjzZV" }, "requestId": "request_8VF0hXMHz0" } }
    , { "ResponseAdded": { "httpStatusCode": 200, "requestId": "request_8VF0hXMHz0", "responseId": "response_Hn5EnUcQOu" } }
    , { "ShapeAdded": { "baseShapeId": "_Todo_0", "name": "", "parameters": { "DynamicParameterList": { "shapeParameterIds": [] } }, "shapeId": "shape_oIoNGLkNiX" } }
    , { "ResponseBodySet": { "bodyDescriptor": { "httpContentType": "application/json; charset=utf-8", "isRemoved": false, "shapeId": "shape_oIoNGLkNiX" }, "responseId": "response_Hn5EnUcQOu" } }
    , { "ContributionAdded": { "id": "request_8VF0hXMHz0", "key": "description", "value": "Create a brand spanking new TODO" } }
    , { "FieldAdded": { "fieldId": "field_MGaT9tKpnz", "name": "zipCode", "shapeDescriptor": { "FieldShapeFromShape": { "fieldId": "field_MGaT9tKpnz", "shapeId": "$number" } }, "shapeId": "_Todo_0" } }
]

describe('flattened shapes', function () {
    const commandContext = new RfcCommandContext('userId', 'sessionId', 'batchId')

    describe('todo example', function () {
        const rfcId = 'test-rfc'
        const eventStore = Facade.makeEventStore();
        eventStore.bulkAdd(rfcId, JSON.stringify(todoEvents));
        const rfcService = Facade.fromJsonCommands(eventStore, rfcId, commandContext, '[]');
        const queries = Queries(eventStore, rfcService, rfcId);
        
        xit('should render ToDo', function () {
            const shapeId = '_Todo_0'
            const output = []
            ShapeUtilities.flatten(queries, shapeId, 0, [], output)
            expect(output).toMatchSnapshot()
        })
        it('should render ToDo', function () {
            const shapeId = '_ResponseBody_0'
            const output = []
            ShapeUtilities.flatten(queries, shapeId, 0, [], output)
            expect(output).toMatchSnapshot()
        })
        
    })
    describe('pagination example', function () {
        const rfcId = 'test-rfc'
        const eventStore = Facade.makeEventStore();
        const rfcService = Facade.fromJsonCommands(eventStore, rfcId, commandContext, '[]')
        const queries = Queries(eventStore, rfcService, rfcId);
        rfcService.handleCommands(rfcId, ...starterCommands)

        it('should render $UserListResponse1', function () {
            const shapeId = '$UserListResponse1'
            const output = []
            ShapeUtilities.flatten(queries, shapeId, 0, [], output)
            expect(output).toMatchSnapshot()
        });
        it('should render $PaginationWrapper', function () {
            const shapeId = '$PaginationWrapper'
            const output = []
            ShapeUtilities.flatten(queries, shapeId, 0, [], output)
            expect(output).toMatchSnapshot()
        });
        it('should render $Account', function () {
            const shapeId = '$Account'
            const output = []
            ShapeUtilities.flatten(queries, shapeId, 0, [], output)
            expect(output).toMatchSnapshot()
        });
    });
});