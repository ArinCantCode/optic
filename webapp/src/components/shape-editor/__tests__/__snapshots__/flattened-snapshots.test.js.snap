// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flattened shapes pagination example should render $Account 1`] = `
Array [
  Object {
    "depth": 0,
    "id": "$Account",
    "isExpandable": true,
    "name": "Account",
    "shapeName": Object {
      "baseShapeName": "Object",
      "bindingInfo": Array [],
      "color": "#32417d",
      "id": "$object",
    },
    "trail": Array [
      "$Account",
    ],
    "type": "shape",
  },
  Object {
    "depth": 1,
    "fieldShapeId": "$AccountId",
    "id": "$Account.id",
    "isExpandable": false,
    "name": "id",
    "parentShapeId": "$Account",
    "shapeName": Object {
      "baseShapeName": "Account ID",
      "bindingInfo": Array [
        Object {
          "binding": Object {
            "ShapeProvider": Object {
              "shapeId": "$Account",
            },
          },
          "boundName": "Account",
          "color": "#32417d",
          "parameterId": "$referenceInner",
          "parameterName": "T",
        },
      ],
      "color": undefined,
      "id": "$AccountId",
    },
    "trail": Array [
      "$Account",
      "$Account.id",
    ],
    "type": "field",
  },
  Object {
    "depth": 1,
    "fieldShapeId": "$list",
    "id": "$Account.userIds",
    "isExpandable": false,
    "name": "userIds",
    "parentShapeId": "$Account",
    "shapeName": Object {
      "baseShapeName": "List",
      "bindingInfo": Array [
        Object {
          "binding": Object {
            "ShapeProvider": Object {
              "shapeId": "$UserId",
            },
          },
          "boundName": "User ID",
          "color": undefined,
          "parameterId": "$listItem",
          "parameterName": "T",
        },
        Object {
          "binding": Object {
            "ShapeProvider": Object {
              "shapeId": "$UserId",
            },
          },
          "boundName": "User ID",
          "color": undefined,
          "parameterId": "$listItem",
          "parameterName": "T",
        },
      ],
      "color": "#7d1e34",
      "id": "$list",
    },
    "trail": Array [
      "$Account",
      "$Account.userIds",
    ],
    "type": "field",
  },
]
`;

exports[`flattened shapes pagination example should render $PaginationWrapper 1`] = `
Array [
  Object {
    "depth": 0,
    "id": "$PaginationWrapper",
    "isExpandable": true,
    "name": "PaginationWrapper",
    "shapeName": Object {
      "baseShapeName": "Object",
      "bindingInfo": Array [],
      "color": "#32417d",
      "id": "$object",
    },
    "trail": Array [
      "$PaginationWrapper",
    ],
    "type": "shape",
  },
  Object {
    "depth": 1,
    "fieldShapeId": "$number",
    "id": "$PaginationWrapper.offset",
    "isExpandable": false,
    "name": "offset",
    "parentShapeId": "$PaginationWrapper",
    "shapeName": Object {
      "baseShapeName": "number",
      "bindingInfo": Array [],
      "color": "#e4508f",
      "id": "$number",
    },
    "trail": Array [
      "$PaginationWrapper",
      "$PaginationWrapper.offset",
    ],
    "type": "field",
  },
  Object {
    "depth": 1,
    "fieldShapeId": "$list",
    "id": "$PaginationWrapper.items",
    "isExpandable": false,
    "name": "items",
    "parentShapeId": "$PaginationWrapper",
    "shapeName": Object {
      "baseShapeName": "List",
      "bindingInfo": Array [
        Object {
          "binding": Object {
            "ParameterProvider": Object {
              "shapeParameterId": "$PaginationWrapper.T",
            },
          },
          "boundName": "T",
          "color": "#229922",
          "parameterId": "$listItem",
          "parameterName": "T",
        },
        Object {
          "binding": Object {
            "ParameterProvider": Object {
              "shapeParameterId": "$PaginationWrapper.T",
            },
          },
          "boundName": "T",
          "color": "#229922",
          "parameterId": "$listItem",
          "parameterName": "T",
        },
      ],
      "color": "#7d1e34",
      "id": "$list",
    },
    "trail": Array [
      "$PaginationWrapper",
      "$PaginationWrapper.items",
    ],
    "type": "field",
  },
]
`;

exports[`flattened shapes pagination example should render $UserListResponse1 1`] = `
Array [
  Object {
    "depth": 0,
    "id": "$UserListResponse1",
    "isExpandable": true,
    "name": "User List Response 1",
    "shapeName": Object {
      "baseShapeName": "PaginationWrapper",
      "bindingInfo": Array [
        Object {
          "binding": Object {
            "ShapeProvider": Object {
              "shapeId": "$User",
            },
          },
          "boundName": "User",
          "color": "#32417d",
          "parameterId": "$PaginationWrapper.T",
          "parameterName": "T",
        },
      ],
      "color": "#32417d",
      "id": "$PaginationWrapper",
    },
    "trail": Array [
      "$UserListResponse1",
    ],
    "type": "shape",
  },
  Object {
    "depth": 1,
    "fieldShapeId": "$number",
    "id": "$PaginationWrapper.offset",
    "isExpandable": false,
    "name": "offset",
    "parentShapeId": "$UserListResponse1",
    "shapeName": Object {
      "baseShapeName": "number",
      "bindingInfo": Array [],
      "color": "#e4508f",
      "id": "$number",
    },
    "trail": Array [
      "$UserListResponse1",
      "$PaginationWrapper.offset",
    ],
    "type": "field",
  },
  Object {
    "depth": 1,
    "fieldShapeId": "$list",
    "id": "$PaginationWrapper.items",
    "isExpandable": false,
    "name": "items",
    "parentShapeId": "$UserListResponse1",
    "shapeName": Object {
      "baseShapeName": "List",
      "bindingInfo": Array [
        Object {
          "binding": Object {
            "ParameterProvider": Object {
              "shapeParameterId": "$PaginationWrapper.T",
            },
          },
          "boundName": null,
          "color": "#229922",
          "parameterId": "$listItem",
          "parameterName": "T",
        },
        Object {
          "binding": Object {
            "ParameterProvider": Object {
              "shapeParameterId": "$PaginationWrapper.T",
            },
          },
          "boundName": null,
          "color": "#229922",
          "parameterId": "$listItem",
          "parameterName": "T",
        },
      ],
      "color": "#7d1e34",
      "id": "$list",
    },
    "trail": Array [
      "$UserListResponse1",
      "$PaginationWrapper.items",
    ],
    "type": "field",
  },
]
`;

exports[`flattened shapes todo example should render ToDo 1`] = `
Array [
  Object {
    "depth": 0,
    "id": "_ResponseBody_0",
    "isExpandable": false,
    "name": "",
    "shapeName": Object {
      "baseShapeName": "List",
      "bindingInfo": Array [
        Object {
          "binding": Object {
            "ShapeProvider": Object {
              "shapeId": "_Todo_0",
            },
          },
          "boundName": "ToDo",
          "color": "#32417d",
          "parameterId": "$listItem",
          "parameterName": "T",
        },
      ],
      "color": "#7d1e34",
      "id": "rG6Eyi_0",
    },
    "trail": Array [
      "_ResponseBody_0",
    ],
    "type": "shape",
  },
]
`;
